File: pixie/evals/dataset_runner.py
"""Dataset-driven test runner for ``pixie test``.

Processes dataset JSON files where each row specifies its own evaluators.
Built-in evaluator names (no dots) are auto-resolved to ``pixie.{Name}``.
Custom evaluators use fully qualified names.

Usage::

    pixie test path/to/dataset.json       # single dataset
    pixie test path/to/dir/               # all datasets in dir tree
    pixie test                            # all datasets in pixie folder
"""

from __future__ import annotations

import importlib
import importlib.util
import inspect
import json
from collections.abc import Callable
from dataclasses import dataclass, field
from json import JSONDecodeError
from pathlib import Path
from typing import Any

from pixie.storage.evaluable import Evaluable

#: Names of all built-in evaluators exported from ``pixie``.
#: When a dataset uses a bare name (no dots), it is looked up here.
BUILTIN_EVALUATOR_NAMES: frozenset[str] = frozenset(
    {
        "LevenshteinMatch",
        "ExactMatch",
        "NumericDiff",
        "JSONDiff",
        "ValidJSON",
        "ListContains",
        "EmbeddingSimilarity",
        "Factuality",
        "ClosedQA",
        "Battle",
        "Humor",
        "Security",
        "Sql",
        "Summary",
        "Translation",
        "Possible",
        "Moderation",
        "ContextRelevancy",
        "Faithfulness",
        "AnswerRelevancy",
        "AnswerCorrectness",
    }
)


def _load_callable(reference: str, base_dir: Path) -> Any:
    """Load a Python object from a ``filepath:name`` reference.

    Uses :func:`importlib.util.spec_from_file_location` to load the
    ``.py`` file directly from disk without any module/package resolution.

    Args:
        reference: Reference in ``filepath:callable_name`` format where
            *filepath* is relative to *base_dir*.
        base_dir: Directory to resolve relative file paths against.

    Returns:
        The resolved Python object (function, class, or instance).

    Raises:
        ValueError: If *reference* does not contain ``:``.
        FileNotFoundError: If the resolved file does not exist.
        ImportError: If the module cannot be loaded from the file.
        AttributeError: If the named attribute does not exist.
    """
    if ":" not in reference:
        raise ValueError(f"Reference must use filepath:name format, got {reference!r}.")
    file_part, attr_name = reference.rsplit(":", 1)
    file_path = (base_dir / file_part).resolve()

    if not file_path.is_file():
        raise FileNotFoundError(f"Module file not found: {file_path}")

    spec = importlib.util.spec_from_file_location("_pixie_user_mod", file_path)
    if spec is None or spec.loader is None:
        raise ImportError(f"Cannot create module spec from {file_path}")

    module = importlib.util.module_from_spec(spec)
    spec.loader.exec_module(module)

    return getattr(module, attr_name)


def resolve_evaluator_name(name: str) -> str:
    """Resolve short built-in name or validate a custom evaluator reference.

    Args:
        name: Either a bare class name (``"Factuality"``) for built-ins
            or a ``filepath:callable_name`` reference for custom evaluators
            (e.g. ``"pixie_qa/evaluators.py:ConciseVoiceStyle"``).

    Returns:
        A resolved evaluator name: ``"pixie.{Name}"`` for built-ins,
        or the original ``filepath:name`` reference for custom evaluators.

    Raises:
        ValueError: If *name* is not a known built-in and does not use
            the ``filepath:name`` format.
    """
    name = name.strip()
    if ":" in name:
        return name
    if name in BUILTIN_EVALUATOR_NAMES:
        return f"pixie.{name}"
    raise ValueError(
        f"Unknown evaluator {name!r}. "
        f"Built-in evaluators use bare names (e.g. 'Factuality'). "
        f"Custom evaluators use filepath:name format "
        f"(e.g. 'pixie_qa/evaluators.py:{name}')."
    )


def _resolve_evaluator(name: str) -> Callable[..., Any]:
    """Import and return an evaluator by name.

    Accepts bare built-in names (``"Factuality"``) and
    ``filepath:callable_name`` references for custom evaluators
    (e.g. ``"pixie_qa/evaluators.py:ConciseVoiceStyle"``).

    The resolved attribute can be:

    - A **class** — instantiated via ``cls()``.
    - A **zero-arg factory function** (e.g. built-in ``ExactMatch``) —
      called to produce the evaluator instance.
    - A **function with required parameters** (evaluator function) —
      returned as-is.
    - A **pre-instantiated callable** (e.g. the return value of
      ``create_llm_evaluator()``) — returned as-is.

    Args:
        name: Evaluator name — bare built-in name or ``filepath:name``.

    Returns:
        A callable evaluator (class instance, function, or closure).

    Raises:
        TypeError: If the resolved attribute is not callable.
    """
    fqn = resolve_evaluator_name(name)

    if ":" in fqn:
        attr = _load_callable(fqn, Path.cwd())
    else:
        module_path, _, attr_name = fqn.rpartition(".")
        module = importlib.import_module(module_path)
        attr = getattr(module, attr_name)

    # Classes are instantiated (e.g. class-based evaluators).
    if isinstance(attr, type):
        instance: Callable[..., Any] = attr()
        return instance

    # Functions: distinguish zero-arg factories (like built-in ExactMatch)
    # from evaluator functions that take evaluable as the first parameter.
    if inspect.isfunction(attr):
        sig = inspect.signature(attr)
        required = [
            p
            for p in sig.parameters.values()
            if p.default is inspect.Parameter.empty
            and p.kind
            not in (inspect.Parameter.VAR_POSITIONAL, inspect.Parameter.VAR_KEYWORD)
        ]
        if not required:
            # Zero-arg factory → call to produce evaluator instance.
            factory_result: Callable[..., Any] = attr()
            return factory_result
        # Has required params → evaluator function itself.
        func_eval: Callable[..., Any] = attr
        return func_eval

    # Callable instances (e.g. create_llm_evaluator result): use as-is.
    if callable(attr):
        callable_eval: Callable[..., Any] = attr
        return callable_eval

    raise TypeError(
        f"Evaluator {name!r} resolved to {type(attr).__name__}, "
        f"which is not callable."
    )


def _resolve_runnable(reference: str) -> Callable[..., Any]:
    """Load a runnable function from a ``filepath:callable_name`` reference.

    Args:
        reference: Reference in ``filepath:callable_name`` format
            (e.g. ``"pixie_qa/scripts/run_app.py:run_app"``).
            The file path is relative to the current working directory.

    Returns:
        The resolved callable.

    Raises:
        ValueError: If *reference* does not use ``filepath:name`` format.
    """
    reference = reference.strip()
    if ":" not in reference:
        raise ValueError(
            f"Runnable must use filepath:name format "
            f"(e.g. 'pixie_qa/scripts/run_app.py:run_app'), "
            f"got {reference!r}."
        )
    func: Callable[..., Any] = _load_callable(reference, Path.cwd())
    return func


async def _noop_runnable(eval_input: object) -> None:
    """No-op runnable for dataset items that already carry ``eval_output``."""


def _short_name(name: str) -> str:
    """Extract the short callable name from a reference.

    Handles both ``filepath:callable_name`` and ``module.ClassName`` formats.
    """
    if ":" in name:
        return name.rsplit(":", 1)[1]
    return name.rpartition(".")[2] or name


def _expand_evaluator_names(
    row_evaluators: list[str] | None,
    default_evaluators: list[str],
) -> list[str]:
    """Resolve row-level evaluator names against defaults.

    Rules:
    - If *row_evaluators* is ``None`` or empty, use *default_evaluators*.
    - ``"..."`` in the row list is replaced with all *default_evaluators*.
    - Otherwise the row list is used as-is.

    Args:
        row_evaluators: Evaluator names declared on the dataset item.
        default_evaluators: Dataset-level default evaluator names.

    Returns:
        The resolved evaluator name list.
    """
    if not row_evaluators:
        return list(default_evaluators)

    result: list[str] = []
    for name in row_evaluators:
        if name.strip() == "...":
            result.extend(default_evaluators)
        else:
            result.append(name)
    return result


@dataclass(frozen=True)
class LoadedDataset:
    """Parsed dataset ready for evaluation.

    Attributes:
        name: Dataset display name.
        runnable: ``filepath:callable_name`` reference for the runnable.
        entries: List of ``(evaluable, evaluator_names)`` pairs.
    """

    name: str
    runnable: str
    entries: list[tuple[Evaluable, list[str]]] = field(default_factory=list)


def discover_dataset_files(path: str) -> list[Path]:
    """Find all dataset JSON files under *path*.

    Args:
        path: A ``.json`` file, a directory, or ``"."`` for current dir.

    Returns:
        Sorted list of dataset file paths.
    """
    target = Path(path)
    if target.is_file() and target.suffix == ".json":
        return [target]
    if target.is_dir():
        return sorted(target.rglob("*.json"))
    return []


def _parse_evaluator_list(
    raw: Any,
    *,
    allow_ellipsis: bool,
    location: str,
    errors: list[str],
) -> list[str]:
    """Parse evaluator list while collecting validation errors."""
    if not isinstance(raw, list):
        errors.append(f"{location}: 'evaluators' must be a list of strings.")
        return []

    names: list[str] = []
    for i, value in enumerate(raw, start=1):
        if not isinstance(value, str) or not value.strip():
            errors.append(f"{location}: evaluator #{i} must be a non-empty string.")
            continue

        name = value.strip()
        if name == "..." and not allow_ellipsis:
            errors.append(f"{location}: '...' is only allowed in row-level evaluators.")
            continue
        names.append(name)

    return names


def _validate_evaluator_names(
    names: list[str],
    *,
    location: str,
    errors: list[str],
) -> None:
    """Validate evaluator names by checking resolution and importability."""
    for name in names:
        try:
            resolve_evaluator_name(name)
            _resolve_evaluator(name)
        except Exception as exc:  # pragma: no cover - exact type varies by import path
            errors.append(
                f"{location}: invalid evaluator {name!r} ({type(exc).__name__}: {exc})."
            )


def validate_dataset_file(dataset_path: Path) -> list[str]:
    """Validate a dataset file and return a list of human-readable errors."""
    if not dataset_path.exists():
        return [f"{dataset_path}: dataset not found."]

    try:
        with open(dataset_path, encoding="utf-8") as f:
            data = json.load(f)
    except JSONDecodeError as exc:
        return [f"{dataset_path}: invalid JSON ({exc.msg} at line {exc.lineno})."]

    errors: list[str] = []
    if not isinstance(data, dict):
        return [f"{dataset_path}: top-level JSON value must be an object."]

    runnable_raw = data.get("runnable")
    if not isinstance(runnable_raw, str) or not runnable_raw.strip():
        errors.append(
            f"{dataset_path}: missing required top-level 'runnable' (non-empty string)."
        )
    else:
        runnable = runnable_raw.strip()
        try:
            resolved = _resolve_runnable(runnable)
            if not callable(resolved):
                errors.append(
                    f"{dataset_path}: runnable {runnable!r} does not resolve to a callable."
                )
        except Exception as exc:  # pragma: no cover - exact type varies by import path
            errors.append(
                f"{dataset_path}: invalid runnable {runnable!r} ({type(exc).__name__}: {exc})."
            )

    default_evaluators_raw = data.get("evaluators", [])
    default_evaluators = _parse_evaluator_list(
        default_evaluators_raw,
        allow_ellipsis=False,
        location=f"{dataset_path} (dataset defaults)",
        errors=errors,
    )
    _validate_evaluator_names(
        default_evaluators,
        location=f"{dataset_path} (dataset defaults)",
        errors=errors,
    )

    items_raw = data.get("items", [])
    if not isinstance(items_raw, list):
        errors.append(f"{dataset_path}: 'items' must be a list.")
        return errors

    for idx, row in enumerate(items_raw, start=1):
        row_location = f"{dataset_path} item #{idx}"
        if not isinstance(row, dict):
            errors.append(f"{row_location}: item must be an object.")
            continue

        description = row.get("description")
        if not isinstance(description, str) or not description.strip():
            errors.append(
                f"{row_location}: missing required 'description' (non-empty string)."
            )

        row_evaluators: list[str] | None = None
        if "evaluators" in row:
            row_evaluators = _parse_evaluator_list(
                row.get("evaluators"),
                allow_ellipsis=True,
                location=row_location,
                errors=errors,
            )

        resolved_evaluators = _expand_evaluator_names(
            row_evaluators, default_evaluators
        )
        resolved_evaluators = [
            name.strip() for name in resolved_evaluators if name.strip()
        ]
        if not resolved_evaluators:
            errors.append(
                f"{row_location}: no evaluators resolved. "
                "Set dataset-level 'evaluators' or row-level 'evaluators'."
            )
            continue

        # Validate concrete resolved names only ("..." is expanded away above).
        _validate_evaluator_names(
            [name for name in resolved_evaluators if name != "..."],
            location=row_location,
            errors=errors,
        )

    return errors


def _is_new_format_item(item_data: dict[str, Any]) -> bool:
    """Return True if the item uses the new entry_input/dependency_input format."""
    return "entry_input" in item_data or "dependency_input" in item_data


def _build_evaluable_new_format(item_data: dict[str, Any]) -> Evaluable:
    """Build an :class:`Evaluable` from a new-format dataset item.

    New format items have ``entry_input`` and optional ``dependency_input``.
    ``dependency_input`` values are jsonpickle-serialized strings keyed by
    wrap name.
    """
    entry_input: Any = item_data.get("entry_input")
    dependency_input_raw = item_data.get("dependency_input") or {}
    dependency_input: dict[str, str] = {}
    for k, v in dependency_input_raw.items():
        if not isinstance(v, str):
            raise ValueError(
                f"dependency_input[{k!r}] must be a jsonpickle-serialized string, "
                f"got {type(v).__name__}. Encode the value with jsonpickle.encode() "
                f"before storing it in the dataset."
            )
        dependency_input[k] = v
    # Build metadata that includes dependency_input for evaluator context.
    meta: dict[str, Any] = {}
    if dependency_input:
        meta["dependency_input"] = dependency_input

    evaluable_data = {
        k: v
        for k, v in item_data.items()
        if k not in ("entry_input", "dependency_input", "eval_input", "eval_metadata")
    }
    evaluable_data["eval_input"] = entry_input
    evaluable_data["eval_metadata"] = meta or None
    return Evaluable.model_validate(evaluable_data)


def load_dataset_entries(
    dataset_path: Path,
) -> LoadedDataset:
    """Load a dataset and return a :class:`LoadedDataset`.

        The dataset JSON requires:

        - top-level ``runnable`` (str) — ``filepath:callable_name`` reference
            to a function that produces ``eval_output`` from ``eval_input``.
        - top-level ``items`` (list[dict]) — dataset entries.
        - per-item ``description`` (str).
        - at least one evaluator per item, via row-level ``evaluators`` or
            dataset-level default ``evaluators``.

        Each item may use either:

        - **New format**: ``entry_input`` (dict) + optional ``dependency_input``
          (dict[str, str] of jsonpickle-serialized values).  The runnable is
          called with ``entry_input`` only; dependency data is injected via
          ``wrap()``.
        - **Legacy format**: ``eval_input`` is passed directly to the runnable,
          and the runnable's return value becomes ``eval_output``.

    Args:
        dataset_path: Path to a dataset JSON file.

    Returns:
        A :class:`LoadedDataset` with resolved entries.

    Raises:
        FileNotFoundError: If *dataset_path* does not exist.
        ValueError: If dataset validation fails.
    """
    if not dataset_path.exists():
        raise FileNotFoundError(f"Dataset not found: {dataset_path}")

    errors = validate_dataset_file(dataset_path)
    if errors:
        message = "Dataset validation failed:\n" + "\n".join(errors)
        raise ValueError(message)

    with open(dataset_path, encoding="utf-8") as f:
        data = json.load(f)

    dataset_name: str = data.get("name", dataset_path.stem)
    runnable = str(data["runnable"]).strip()
    default_evaluators_raw: list[str] = data.get("evaluators", [])
    default_evaluators = [
        n.strip() for n in default_evaluators_raw if isinstance(n, str) and n.strip()
    ]
    raw_items: list[dict[str, Any]] = data.get("items", [])

    entries: list[tuple[Evaluable, list[str]]] = []
    for item_data in raw_items:
        row_evaluators_raw = item_data.get("evaluators")
        if isinstance(row_evaluators_raw, list):
            row_evaluators = [n for n in row_evaluators_raw if isinstance(n, str)]
        else:
            row_evaluators = None

        evaluator_names = _expand_evaluator_names(row_evaluators, default_evaluators)
        evaluator_names = [n.strip() for n in evaluator_names if n.strip()]

        if _is_new_format_item(item_data):
            evaluable = _build_evaluable_new_format(item_data)
        else:
            evaluable = Evaluable.model_validate(item_data)

        entries.append((evaluable, evaluator_names))

    return LoadedDataset(
        name=dataset_name,
        runnable=runnable,
        entries=entries,
    )

File: ts/src/evals/datasetRunner.ts
/**
 * Dataset-driven test runner for `pixie test`.
 *
 * Processes dataset JSON files where each row specifies its own evaluators.
 * Built-in evaluator names (no dots) are auto-resolved to `pixie-qa.{Name}`.
 * Custom evaluators use fully qualified names.
 */

import fs from "fs";
import path from "path";

import type { Evaluable, JsonValue } from "../storage/evaluable";
import { UNSET } from "../storage/evaluable";

// ── Constants ────────────────────────────────────────────────────────────────

/** Names of all built-in evaluators. */
export const BUILTIN_EVALUATOR_NAMES: ReadonlySet<string> = new Set([
  "LevenshteinMatch",
  "ExactMatch",
  "NumericDiff",
  "JSONDiff",
  "ValidJSON",
  "ListContains",
  "EmbeddingSimilarity",
  "Factuality",
  "ClosedQA",
  "Battle",
  "Humor",
  "Security",
  "Sql",
  "Summary",
  "Translation",
  "Possible",
  "Moderation",
  "ContextRelevancy",
  "Faithfulness",
  "AnswerRelevancy",
  "AnswerCorrectness",
]);

// ── Types ────────────────────────────────────────────────────────────────────

/** Parsed dataset ready for evaluation. */
export interface LoadedDataset {
  readonly name: string;
  /** Fully qualified name of the runnable function. */
  readonly runnable: string;
  /** List of [evaluable, evaluatorNames] pairs. */
  readonly entries: Array<[Evaluable, string[]]>;
}

// ── Functions ────────────────────────────────────────────────────────────────

/**
 * Resolve short built-in name to fully qualified, or pass through FQN.
 *
 * @throws if name has no dots and is not a known built-in.
 */
export function resolveEvaluatorName(name: string): string {
  const trimmed = name.trim();
  if (trimmed.includes(".")) return trimmed;
  if (BUILTIN_EVALUATOR_NAMES.has(trimmed)) {
    return `pixie-qa.${trimmed}`;
  }
  throw new Error(
    `Unknown evaluator ${JSON.stringify(trimmed)}. ` +
      `Use a fully qualified name for custom evaluators ` +
      `(e.g. 'myapp.evals.${trimmed}').`
  );
}

/**
 * Import and instantiate an evaluator by name.
 */
function resolveEvaluator(name: string): unknown {
  const fqn = resolveEvaluatorName(name);
  const lastDot = fqn.lastIndexOf(".");
  const modulePath = fqn.substring(0, lastDot);
  const className = fqn.substring(lastDot + 1);

  // eslint-disable-next-line @typescript-eslint/no-require-imports
  const mod = require(modulePath);
  const cls = mod[className];
  if (typeof cls === "function") {
    return cls();
  }
  return cls;
}

/**
 * Import a runnable function by fully qualified name.
 */
function resolveRunnable(fqn: string): unknown {
  const lastDot = fqn.lastIndexOf(".");
  if (lastDot === -1) {
    throw new Error(
      `Runnable must be a fully qualified name (e.g. 'myapp.module.func'), ` +
        `got ${JSON.stringify(fqn)}.`
    );
  }
  const modulePath = fqn.substring(0, lastDot);
  const funcName = fqn.substring(lastDot + 1);

  // eslint-disable-next-line @typescript-eslint/no-require-imports
  const mod = require(modulePath);
  return mod[funcName];
}

/** Extract the class name from a possibly fully qualified name. */
function shortName(name: string): string {
  const lastDot = name.lastIndexOf(".");
  return lastDot === -1 ? name : name.substring(lastDot + 1);
}

/**
 * Resolve row-level evaluator names against defaults.
 *
 * - If `rowEvaluators` is null or empty, use `defaultEvaluators`.
 * - `"..."` in the row list is replaced with all `defaultEvaluators`.
 */
function expandEvaluatorNames(
  rowEvaluators: string[] | null,
  defaultEvaluators: string[]
): string[] {
  if (!rowEvaluators || rowEvaluators.length === 0) {
    return [...defaultEvaluators];
  }

  const result: string[] = [];
  for (const name of rowEvaluators) {
    if (name.trim() === "...") {
      result.push(...defaultEvaluators);
    } else {
      result.push(name);
    }
  }
  return result;
}

/** Return sorted list of all built-in evaluator names. */
export function listAvailableEvaluators(): string[] {
  return [...BUILTIN_EVALUATOR_NAMES].sort();
}

/**
 * Find all dataset JSON files under `searchPath`.
 *
 * Handles file, directory, or `.` for current dir.
 */
export function discoverDatasetFiles(searchPath: string): string[] {
  const target = path.resolve(searchPath);
  const stat = fs.statSync(target, { throwIfNoEntry: false });

  if (stat?.isFile() && target.endsWith(".json")) {
    return [target];
  }

  if (stat?.isDirectory()) {
    return findJsonFiles(target).sort();
  }

  return [];
}

function findJsonFiles(dir: string): string[] {
  const results: string[] = [];
  const entries = fs.readdirSync(dir, { withFileTypes: true });
  for (const entry of entries) {
    const fullPath = path.join(dir, entry.name);
    if (entry.isDirectory()) {
      results.push(...findJsonFiles(fullPath));
    } else if (entry.isFile() && entry.name.endsWith(".json")) {
      results.push(fullPath);
    }
  }
  return results;
}

// ── Validation helpers ───────────────────────────────────────────────────────

function parseEvaluatorList(
  raw: unknown,
  opts: {
    allowEllipsis: boolean;
    location: string;
    errors: string[];
  }
): string[] {
  if (!Array.isArray(raw)) {
    opts.errors.push(
      `${opts.location}: 'evaluators' must be a list of strings.`
    );
    return [];
  }

  const names: string[] = [];
  for (let i = 0; i < raw.length; i++) {
    const value = raw[i];
    if (typeof value !== "string" || !value.trim()) {
      opts.errors.push(
        `${opts.location}: evaluator #${i + 1} must be a non-empty string.`
      );
      continue;
    }
    const name = value.trim();
    if (name === "..." && !opts.allowEllipsis) {
      opts.errors.push(
        `${opts.location}: '...' is only allowed in row-level evaluators.`
      );
      continue;
    }
    names.push(name);
  }
  return names;
}

function validateEvaluatorNames(
  names: string[],
  opts: { location: string; errors: string[] }
): void {
  for (const name of names) {
    try {
      resolveEvaluatorName(name);
      resolveEvaluator(name);
    } catch (exc) {
      const error = exc as Error;
      opts.errors.push(
        `${opts.location}: invalid evaluator ${JSON.stringify(name)} (${error.name}: ${error.message}).`
      );
    }
  }
}

/**
 * Validate a dataset file and return a list of human-readable errors.
 * Empty list means the file is valid.
 */
export function validateDatasetFile(datasetPath: string): string[] {
  if (!fs.existsSync(datasetPath)) {
    return [`${datasetPath}: dataset not found.`];
  }

  let data: unknown;
  try {
    const raw = fs.readFileSync(datasetPath, "utf-8");
    data = JSON.parse(raw);
  } catch (exc) {
    const error = exc as SyntaxError;
    return [`${datasetPath}: invalid JSON (${error.message}).`];
  }

  if (typeof data !== "object" || data === null || Array.isArray(data)) {
    return [`${datasetPath}: top-level JSON value must be an object.`];
  }

  const errors: string[] = [];
  const obj = data as Record<string, unknown>;

  const runnableRaw = obj["runnable"];
  if (typeof runnableRaw !== "string" || !runnableRaw.trim()) {
    errors.push(
      `${datasetPath}: missing required top-level 'runnable' (non-empty string).`
    );
  } else {
    const runnable = runnableRaw.trim();
    try {
      const resolved = resolveRunnable(runnable);
      if (typeof resolved !== "function") {
        errors.push(
          `${datasetPath}: runnable ${JSON.stringify(runnable)} does not resolve to a callable.`
        );
      }
    } catch (exc) {
      const error = exc as Error;
      errors.push(
        `${datasetPath}: invalid runnable ${JSON.stringify(runnable)} (${error.name}: ${error.message}).`
      );
    }
  }

  const defaultEvaluatorsRaw = obj["evaluators"] ?? [];
  const defaultEvaluators = parseEvaluatorList(defaultEvaluatorsRaw, {
    allowEllipsis: false,
    location: `${datasetPath} (dataset defaults)`,
    errors,
  });
  validateEvaluatorNames(defaultEvaluators, {
    location: `${datasetPath} (dataset defaults)`,
    errors,
  });

  const itemsRaw = obj["items"] ?? [];
  if (!Array.isArray(itemsRaw)) {
    errors.push(`${datasetPath}: 'items' must be a list.`);
    return errors;
  }

  for (let idx = 0; idx < itemsRaw.length; idx++) {
    const row = itemsRaw[idx];
    const rowLocation = `${datasetPath} item #${idx + 1}`;

    if (typeof row !== "object" || row === null || Array.isArray(row)) {
      errors.push(`${rowLocation}: item must be an object.`);
      continue;
    }

    const rowObj = row as Record<string, unknown>;
    const description = rowObj["description"];
    if (typeof description !== "string" || !description.trim()) {
      errors.push(
        `${rowLocation}: missing required 'description' (non-empty string).`
      );
    }

    let rowEvaluators: string[] | null = null;
    if ("evaluators" in rowObj) {
      rowEvaluators = parseEvaluatorList(rowObj["evaluators"], {
        allowEllipsis: true,
        location: rowLocation,
        errors,
      });
    }

    let resolvedEvaluators = expandEvaluatorNames(
      rowEvaluators,
      defaultEvaluators
    );
    resolvedEvaluators = resolvedEvaluators
      .map((n) => n.trim())
      .filter((n) => n.length > 0);

    if (resolvedEvaluators.length === 0) {
      errors.push(
        `${rowLocation}: no evaluators resolved. ` +
          "Set dataset-level 'evaluators' or row-level 'evaluators'."
      );
      continue;
    }

    validateEvaluatorNames(
      resolvedEvaluators.filter((n) => n !== "..."),
      { location: rowLocation, errors }
    );
  }

  return errors;
}

/**
 * Load a dataset and return a LoadedDataset.
 *
 * @throws if the file does not exist or validation fails.
 */
export function loadDatasetEntries(datasetPath: string): LoadedDataset {
  if (!fs.existsSync(datasetPath)) {
    throw new Error(`Dataset not found: ${datasetPath}`);
  }

  const validationErrors = validateDatasetFile(datasetPath);
  if (validationErrors.length > 0) {
    const message =
      "Dataset validation failed:\n" + validationErrors.join("\n");
    throw new Error(message);
  }

  const raw = JSON.parse(fs.readFileSync(datasetPath, "utf-8")) as Record<
    string,
    unknown
  >;
  const datasetName = (raw["name"] as string) ?? path.basename(datasetPath, ".json");
  const runnable = String(raw["runnable"]).trim();
  const defaultEvaluatorsRaw = (raw["evaluators"] ?? []) as unknown[];
  const defaultEvaluators = defaultEvaluatorsRaw
    .filter((n): n is string => typeof n === "string" && n.trim().length > 0)
    .map((n) => n.trim());

  const rawItems = (raw["items"] ?? []) as Array<Record<string, unknown>>;

  const entries: Array<[Evaluable, string[]]> = [];
  for (const itemData of rawItems) {
    const rowEvaluatorsRaw = itemData["evaluators"];
    let rowEvaluators: string[] | null = null;
    if (Array.isArray(rowEvaluatorsRaw)) {
      rowEvaluators = rowEvaluatorsRaw.filter(
        (n): n is string => typeof n === "string"
      );
    }

    let evaluatorNames = expandEvaluatorNames(
      rowEvaluators,
      defaultEvaluators
    );
    evaluatorNames = evaluatorNames
      .map((n) => n.trim())
      .filter((n) => n.length > 0);

    const evaluable: Evaluable = {
      evalInput: (itemData["evalInput"] ?? itemData["eval_input"] ?? null) as JsonValue,
      evalOutput: (itemData["evalOutput"] ?? itemData["eval_output"] ?? null) as JsonValue,
      evalMetadata:
        (itemData["evalMetadata"] ??
          itemData["eval_metadata"] ??
          null) as Record<string, JsonValue> | null,
      expectedOutput:
        itemData["expectedOutput"] !== undefined
          ? (itemData["expectedOutput"] as JsonValue)
          : itemData["expected_output"] !== undefined
            ? (itemData["expected_output"] as JsonValue)
            : UNSET,
      evaluators: (itemData["evaluators"] as readonly string[]) ?? null,
      description: (itemData["description"] as string) ?? null,
    };

    entries.push([evaluable, evaluatorNames]);
  }

  return {
    name: datasetName,
    runnable,
    entries,
  };
}
